README.md
# Technosoftware GmbH OPC UA .NET Standard based SDKs and Tools

Technosoftware GmbH offers the following .NET Standard OPC UA based Products:
 * [OPC UA Client SDK .NET Standard](https://technosoftware.com/opc-ua-client-sdk-net-standard/)
 * [OPC UA Server SDK .NET Standard](https://technosoftware.com/opc-ua-server-sdk-net-standard/)

## OPC UA Client SDK .NET Standard

The OPC UA Client SDK .NET Standard offers a fast and easy access to the OPC UA Client technology. Develop OPC UA 1.00, 1.01, 1.02, 1.03 and 1.04 compliant UA Clients with C#/VB.NET targeting the .NET Standard.

.Net Standard allows you develop applications that run on all common platforms available today, including Linux, iOS, Android (via Xamarin) and Windows 7/8.1/10 (including embedded/IoT editions) without requiring platform-specific modifications.

The OPC UA Client SDK .NET API defines classes which can be used to implement an OPC client capable to access OPC servers supporting different profiles with the same API. These classes manage client side state information; provide higher level abstractions for OPC tasks such as managing sessions and subscriptions or saving and restoring connection information for later use.

## OPC UA Server SDK .NET Standard

The OPC UA Server SDK .NET Standard offers a fast and easy access to the OPC Unified Architecture (UA) technology. Develop OPC UA 1.00, 1.01, 1.02, 1.03  and 1.04 compliant Servers with C#/VB.NET or any other compiler capable of generating a .NET assembly.

.NET Standard allows you develop applications that run on all common platforms available today, including Linux, iOS, Android (via Xamarin) and Windows 7/8.1/10 (including embedded/IoT editions) without requiring platform-specific modifications.

The developer can concentrate on his application and servers can be developed fast and easily without the need to spend a lot of time learning how to implement the OPC Unified Architecture specification. The server API is easy to use and many OPC specific functions are handled by the framework.

The included OPC Foundation Model Compiler can be used to create the necessary C# classes of Information Model’s specified in XML and CSV based files. At the moment the XML files must be edited by a text editor. 

Documentation of the Mode Compiler can be found at https://github.com/OPCFoundation/UA-ModelCompiler

## Getting started

You can use the following documents as a starting point:

1. [OPC UA NET Standard Installation Guide](/doc/pdf/OPC_UA_NET_Standard_Installation_Guide.pdf)
2. [OPC UA SDKs NET Standard Introduction](/doc/pdf/OPC_UA_SDKs_NET_Standard_Introduction.pdf)
3. [OPC UA Client Development with NET Standard](/doc/pdf/OPC_UA_Client_Development_with_NET_Standard.pdf)
4. [OPC UA Server Development with NET Standard](/doc/pdf/OPC_UA_Server_Development_with_NET_Standard.pdf)

## Workshop

You can also use the PDFs provided from our Workshop available in the [here](/Workshop).

# Evaluation Usage

Evaluation versions are NOT intended to be used for product development! They may differ in features from the product versions and might need some time to get integrated. They are intended for evaluation purposes only.

Product development should always be done on product versions of the SDKs! 

Without a Product License (see below) the contents of this repository is used as an evaluation version.

Evaluation Editions comes with the following restrictions:

 * Evaluation time limited to 30 days and limited to 90 minutes run-time per application start
 * Some evaluations may also show a splash screen at startup

## Product License

If you purchased a product license (Profeessional or Ultimate) of one of the following products

 * [OPC UA Client SDK .NET Standard](https://technosoftware.com/opc-ua-client-sdk-net-standard/)
 * [OPC UA Server SDK .NET Standard](https://technosoftware.com/opc-ua-server-sdk-net-standard/)

you can use the delivered license to turn the contents of this repository into a product version.

# Commercial Support
Purchase commercial suppourt via https://technosoftware.com/product/standard-support/ for a single Standard Support request (two hours support time budget, five business days maximum initial response time, although we usually answer within one to two business days) for EUR 390,00 from Technosoftware GmbH.

By purchasing support, you agree to our Support Service Agreement at https://technosoftware.com/documents/Support_Services_Agreement.pdf

CHANGELOG.md
-------------------------------------------------------------------------------------------------------------
## OPC UA Client SDK .NET Standard - 1.2.1130 (Release Date 30-NOV-2019)

###	Enhancement
- Description and RolePermissions are now initialized

## OPC UA Server SDK .NET Standard - 1.2.1130 (Release Date 30-NOV-2019)

###	Enhancement
- Description and RolePermissions are now initialized

## OPC UA Core .NET Standard - 1.2.1130 (Release Date 30-NOV-2019)

###	Enhancement
- Description and RolePermissions are now initialized
- Description for OPC UA standard nodes are now loaded also.

-------------------------------------------------------------------------------------------------------------
## OPC UA Client SDK .NET Standard - 1.2.1125 (Release Date 25-NOV-2019)

###	Enhancement
- Less logging for Info messages. Moved them to OperationDetail

## OPC UA Server SDK .NET Standard - 1.2.1125 (Release Date 25-NOV-2019)

###	Enhancement
- Less logging for Info messages. Moved them to OperationDetail

## OPC UA Core .NET Standard - 1.2.1125 (Release Date 25-NOV-2019)

###	Enhancement
- Less logging for Info messages. Moved them to OperationDetail

-------------------------------------------------------------------------------------------------------------
## OPC UA Client SDK .NET Standard - 1.2.1124 (Release Date 24-NOV-2019)

###	Issues
- Add server nonce validation in Client session.

## OPC UA Server SDK .NET Standard - 1.2.1124 (Release Date 24-NOV-2019)

###	Issues
- Added user authentication handling for UaMonitoredItem.

## OPC UA Core .NET Standard - 1.2.1124 (Release Date 24-NOV-2019)

###	Enhancement
- Preparation for Complex Types

###	Issues
- Avoid deep recursion in TcpMessageSocket.ReadNextBlock()
- Certificate Validation fix.
- Remove BadCertificateUriInvalid from list of certificate validation errors that may be suppressed.
- Add server nonce validation in Client session.
- Length validation in user token decryption.
- Stability in BinaryDecoder improved.
- Fix in DiscoveryClient.Create()

-------------------------------------------------------------------------------------------------------------

## OPC UA Server SDK .NET Standard - 1.2.1101 (Release Date 01-NOV-2019)

###	Breaking Changes
- Changed license handling. You will get some errors using this version now. Please see the new license certificate. The new method to use is 
        void OnGetLicenseInformation(out Opc.Ua.LicenseHandler.LicenseEdition productEdition, out string serialNumber);

###	Enhancement
- Added CreateBaseDataVariable(NodeState parent, string path, string name, ExpandedNodeId expandedNodeId, int valueRank, byte accessLevel, object initialValue) method to IUaServer

## OPC UA Client SDK .NET Standard - 1.2.1101 (Release Date 01-NOV-2019)

###	Breaking Changes
- Changed license handling. You will get some errors using this version now. Please see the new license certificate. The new method to use is 
        public static bool Validate(Opc.Ua.LicenseHandler.LicenseEdition productEdition, string serialNumber)

###	Issues
- Add check for some configuration fields and return some additional errors.

## OPC UA Core .NET Standard - 1.2.1101 (Release Date 01-NOV-2019)

###	Issues
- ExpandedNodeId.CompareTo contract violation
- fix schema validator and dictionary loader (only binary)
- fix schema validator and dictionary loader (only binary scheme)
- include some .bsd files for validation

-------------------------------------------------------------------------------------------------------------

## OPC UA Server SDK .NET Standard - 1.2.1013 (Release Date 13-OCT-2019)

###	Breaking Changes
- Several synchronous methods marked as obsolete in 1.1 are now giving an error while building with 1.2. Use the asynchronous versions mentioned in the error instead.
- Updated .NET 4.6.1 version to .NET 4.6.2
  
###	Enhancement
- Add IsNodeAccessibleForUser() and IsReferenceAccessibleForUser() to UaBaseNodeManager().
  This allows to hide nodes and references for specific users. 
  The ServerForms and ServerConsole sample shows the usage if these new methods.

###	Issues
- Added several XSD schema's to the schema folder. The Workshop server sample ModelDesign.xml files references the UAModelDesign.xsd. 
  Visual Studio can be used to change the Design files while giving context sensitive and popups. 

## OPC UA Client SDK .NET Standard - 1.2.1013 (Release Date 13-OCT-2019)

###	Breaking Changes
- Several synchronous methods marked as obsolete in 1.1 are now giving an error while building with 1.2. Use the asynchronous versions mentioned in the error instead.
- Updated .NET 4.6.1 version to .NET 4.6.2
- OPC UA Client Gateway is no longer maintained and source code is no longer delivered.

###	Issues
- The Session Method "ReadValue" delivers old values. 
- Using IUaServerData.Status can be made thread safe by using Status.Lock
- Client keep-alive fix - Outstanding publish requests should not stop keep alive reads.

## OPC UA Core .NET Standard - 1.2.1013 (Release Date 13-OCT-2019)

###	Enhancement
- Synchronize with newest UA Nodeset

###	Issues
- TraceEventHandler event does not publish exception information
- Possible handle leak in stress test for TCP communication 
- Interoperability fixes for HTTPS
- Improvement to enable UANodeSet to import/export AccessLevelEx as uint32

-------------------------------------------------------------------------------------------------------------

## OPC UA Server SDK .NET Standard - 1.1.815 (Release Date 15-AUG-2019)

###	Changes
Version numbering scheme changed. Version numbers are used with Major.Minor.Build, where
- Major: is incremented when we make incompatible API changes
- Minor: is incremented when we add functionality in a backwards-compatible manner
- Build: is incremented when we make backwards-compatible bug fixes and is the date (mmyy) the build was triggered

###	Updates / Fixes
- Enhanced documentation and added example usage on Linux, macOS

examples/Workshop/ServerForms/UaServerPlugin.cs
#region Copyright (c) 2011-2019 Technosoftware GmbH. All rights reserved
//-----------------------------------------------------------------------------
// Copyright (c) 2011-2019 Technosoftware GmbH. All rights reserved
// Web: https://technosoftware.com 
// 
// License: 
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//
// SPDX-License-Identifier: MIT
//-----------------------------------------------------------------------------
#endregion Copyright (c) 2011-2019 Technosoftware GmbH. All rights reserved

#region Using Directives

using System;
using System.Collections.Generic;
using System.Reflection;
using System.Globalization;
using System.IO;
using System.Windows.Forms;

using Opc.Ua;

using Technosoftware.UaServer;
using Technosoftware.UaServer.Controls;
using Technosoftware.WorkshopServerForms.Model;

#endregion

namespace Technosoftware.WorkshopServerForms
{

    /// <summary>
    /// OPC Server Configuration and IO Handling
    ///
    /// This C# based plugin for the OPC UA Server SDK .NET shows a base 
    /// OPC UA 1.01 server implementation. At startup items with several 
    /// data types and access rights are statically defined. 
    /// </summary>
    public class UaServerPlugin : IUaServerPlugin, IUaOptionalServerPlugin, IDisposable
    {
        #region Fields

        private IUaServer opcServer_;

        // Track whether Dispose has been called.
        private bool disposed_;
        private readonly object lockDisposable_ = new object();

        #endregion

        #region Constructors, Destructor, Initialization

        // Use C# destructor syntax for finalization code.
        // This destructor will run only if the Dispose method
        // does not get called.
        // It gives your base class the opportunity to finalize.
        // Do not provide destructors in types derived from this class.
        ~UaServerPlugin()
        {
            // Do not re-create Dispose clean-up code here.
            // Calling Dispose(false) is optimal in terms of
            // readability and maintainability.
            Dispose(false);
        }

        /// <summary>
        /// Implement IDisposable.
        /// Do not make this method virtual.
        /// A derived class should not be able to override this method.
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            // This object will be cleaned up by the Dispose method.
            // Therefore, you should call GC.SuppressFinalize to
            // take this object off the finalization queue
            // and prevent finalization code for this object
            // from executing a second time.
            GC.SuppressFinalize(this);
        }

        /// <summary>
        /// Dispose(bool disposing) executes in two distinct scenarios.
        /// If disposing equals true, the method has been called directly
        /// or indirectly by a user's code. Managed and unmanaged resources
        /// can be disposed.
        /// If disposing equals false, the method has been called by the
        /// runtime from inside the finalizer and you should not reference
        /// other objects. Only unmanaged resources can be disposed.
        /// </summary>
        /// <param name="disposing">If disposing equals true, the method has been called directly
        /// or indirectly by a user's code. Managed and unmanaged resources
        /// can be disposed.
        /// If disposing equals false, the method has been called by the
        /// runtime from inside the finalizer and you should not reference
        /// other objects. Only unmanaged resources can be disposed.</param>
        protected virtual void Dispose(bool disposing)
        {
            // Check to see if Dispose has already been called.
            if (!disposed_)
            {
                lock (lockDisposable_)
                {
                    // If disposing equals true, dispose all managed
                    // and unmanaged resources.
                    if (disposing)
                    {
                    }

                    // Call the appropriate methods to clean up
                    // unmanaged resources here.
                    // If disposing is false,
                    // only the following code is executed.

                    // Disposing has been done.
                    disposed_ = true;
                }

            }
        }

        #endregion

        #region General Methods (not related to an OPC specification)

        //---------------------------------------------------------------------
        // OPC UA Server SDK .NET Interface 
        // (Called by the generic server)
        //---------------------------------------------------------------------

        /// <summary>
        /// This method is called from the generic server to get the license information. 
        /// </summary>
        /// <param name="productEdition">Product Edition</param>
        /// <param name="serialNumber">Serial Number</param>
        /// <remarks>Returning empty strings activates the evaluation version of the OPC UA Server SDK .NET. The evaluation allows the usage of the full product for 30 days.</remarks>
        public void OnGetLicenseInformation(out Opc.Ua.LicenseHandler.LicenseEdition productEdition, out string serialNumber)
        {
            productEdition = Opc.Ua.LicenseHandler.LicenseEdition.Evaluation;
            serialNumber = "";
        }

        /// <summary>This method is the first method called from the generic server at the startup.</summary>
        /// <param name="args">String array with the command line parameters as they were specified when the server was being started.</param>
        /// <remarks>
        /// 	<para>The following command line parameters are handled by the generic server:</para>
        /// 	<list type="bullet">
        /// 		<item>/silent<br/>
        ///     No output is done during startup of the server</item>
        /// 		<item>/configfile<br/>
        ///     Allows the definition of a specific application configuration file</item>
        /// 		<item>/install<br/>
        ///     Installs and configures the application according the configuration file, e.g. certificates are created and firewall configured.</item>
        /// 		<item>/uninstall<br/>
        ///     Uninstalls the application by removing the changes made during installation.</item>
        /// 	</list>
        /// </remarks>
        /// <returns>A <see cref="StatusCode"/> code with the result of the operation. Returning an error code stops the further server execution.</returns>
        public StatusCode OnStartup(string[] args)
        {
            return StatusCodes.Good;
        }

        /// <summary>
        /// Defines namespaces used by the application.
        /// </summary>
        /// <returns>Array of namespaces that are used by the application.</returns>
        public string[] OnGetNamespaceUris()
        {
            // set one namespace for the type model.
            var namespaceUrls = new string[3];
            namespaceUrls[0] = Technosoftware.WorkshopServerForms.Model.Namespaces.WorkshopServerForms;
            namespaceUrls[1] = Technosoftware.WorkshopServerForms.Model.Namespaces.Engineering;
            namespaceUrls[2] = Technosoftware.WorkshopServerForms.Model.Namespaces.Operations;
            return namespaceUrls;
        }

        /// <summary>This method is called after the node manager is initialized.</summary>
        /// <param name="opcServer">The generic server object. Used to call methods the generic server provides.</param>
        /// <param name="configuration">The application configuration</param>
        public void OnInitialized(IUaServer opcServer, ApplicationConfiguration configuration)
        {
            opcServer_ = opcServer;
        }

        /// <summary>
        /// 	<para>
        ///         This method is called from the generic server at the startup; when the first client connects or the service is started. All items supported by the server need to be defined by calling the methods provided by the <see cref="IUaServer">IUaServer</see> interface for each item.
        ///     </para>
        /// </summary>
        /// <param name="externalReferences">The externalReferences allows the generic server to link to the general nodes.</param>
        /// <returns>The root folder.</returns>
        public NodeState OnCreateAddressSpace(IDictionary<NodeId, IList<IReference>> externalReferences)
        {
            // Not called because the method CreateAddressSpace() is overwritten in the WorkshopServerNodeManager class
            return null;
        }

        /// <summary>
        /// This method is called from the generic server when the server was successfully started and is running. 
        /// </summary>
        /// <remarks>This method should only return if the server should be stopped.</remarks>
        /// <returns>A <see cref="StatusCode"/> code with the result of the operation. Returning from this method stops the further server execution.</returns>
        public StatusCode OnRunning()
        {
            // Initialize the user interface.
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);

            // run the application interactively.
            Application.Run(new ServerForm(opcServer_));

            return StatusCodes.Good;
        }

        /// <summary>
        /// This method is called from the generic server when a Shutdown is executed.
        /// To ensure proper process shutdown, any communication channels should be closed and all threads terminated before this method returns.
        /// </summary>
        /// <param name="serverState">The current state of the server.</param>
        /// <param name="reason">The reason why the server shutdowns.</param>
        /// <param name="exception">The exception which caused an error. null if not an exception.</param>
        /// <returns>A <see cref="StatusCode"/> code with the result of the operation.</returns>
        public StatusCode OnShutdown(ServerState serverState, string reason, Exception exception)
        {
            return StatusCodes.Good;
        }

        /// <summary>The OPC Server Properties of the current server instance.</summary>
        /// <returns>A <see cref="ServerProperties"/> object.</returns>
        public ServerProperties OnGetServerProperties()
        {
            var properties = new ServerProperties
                                {
                                    ManufacturerName = "Technosoftware GmbH",
                                    ProductName = "Technosoftware OPC UA Workshop Server Sample",
                                    ProductUri = "http://technosoftware.com/WorkshopServer/v1.0",
                                    SoftwareVersion = GetAssemblySoftwareVersion(),
                                    BuildNumber = GetAssemblyBuildNumber(),
                                    BuildDate = GetAssemblyTimestamp()
                                };

            return properties;
        }

        #endregion

        #region Core Server Facet Methods

        //---------------------------------------------------------------------
        // OPC UA Server SDK .NET Interface 
        // (Called by the generic server)
        //---------------------------------------------------------------------

        /// <summary>This method is called when a client executes a 'write' server call. The <see cref="BaseDataVariableState"/> includes all information required to identify the item
        /// as well as original (unmodified) value, timestamp and quality. Depending on the returned result code the cache is updated or not in the generic server after returning from this method.</summary>
        /// <param name="originalVariableState">The <see cref="BaseVariableState"/> of the variable including the original state of the variable. Can be used to check what changes happens</param>
        /// <param name="value">The value which should be written. The returned value is used for updating the cache depending on the returned result code.</param>
        /// <param name="statusCode">A <see cref="StatusCode"/> code which should be used as new status code for the value.</param>
        /// <param name="timestamp">The timestamp the value was written. The returned value is used for updating the cache depending on the returned result code.</param>
        /// <returns>A <see cref="StatusCode"/> code with the result of the operation.</returns>
        /// <remarks>
        /// 	<para>Rules for updating the cache:</para>
        /// 	<list type="number">
        /// 		<item>If the returned <see cref="StatusCode"/> is Bad (something like Bad...) the cache is not updated with timestamp and value.</item>
        /// 		<item>If the returned <see cref="StatusCode"/> is <see cref="StatusCodes.GoodCompletesAsynchronously">OpcStatusCodes.GoodCompletesAsynchronously</see> the cache is not updated with
        ///     timestamp and value. After the customization DLL has finished its operation it can use <see cref="IUaServer.WriteBaseVariable">WriteBaseVariable</see> to update the cache.</item>
        /// 		<item>In all other cases the cache is updated with timestamp and value.</item>
        /// 		<item>In all cases the status code is updated with the status code set in the 'statusCode' parameter.</item>
        /// 	</list>
        /// </remarks>
        public StatusCode OnWriteBaseVariable(BaseVariableState originalVariableState, ref object value, ref StatusCode statusCode, ref DateTime timestamp)
        {
            // V1.3.0.0: To be compliant the provided timestamp must be used. So only if the provided timestamp is invalid set a correct one.
            if (timestamp == DateTime.MinValue)
            {
                timestamp = DateTime.UtcNow;
            }
            return StatusCodes.Good;
        }

        #endregion

        #region Optional Server Plugin methods

        public UaBaseServer OnGetServer()
        {
            return new WorkshopServerForms();
        }

        public UaBaseNodeManager OnGetNodeManager(IUaServer opcServer, IUaServerData uaServer, ApplicationConfiguration configuration, params string[] namespaceUris)
        {
            return new WorkshopServerFormsNodeManager(opcServer, this, uaServer, configuration, namespaceUris);
        }

        #endregion

        #region Helper Methods

        /// <summary>
        /// Returns the linker timestamp for an assembly. 
        /// </summary>
        private static DateTime GetAssemblyTimestamp()
        {
            const int peHeaderOffset = 60;
            const int linkerTimestampOffset = 8;

            var bytes = new byte[2048];

            using (Stream istrm = new FileStream(Assembly.GetCallingAssembly().Location, FileMode.Open, FileAccess.Read))
            {
                istrm.Read(bytes, 0, bytes.Length);
            }

            // get the location of te PE header.
            var index = BitConverter.ToInt32(bytes, peHeaderOffset);

            // get the timestamp from the linker.
            var secondsSince1970 = BitConverter.ToInt32(bytes, index + linkerTimestampOffset);

            // convert to DateTime value.
            var timestamp = new DateTime(1970, 1, 1, 0, 0, 0);
            timestamp = timestamp.AddSeconds(secondsSince1970);
            return timestamp;
        }

        /// <summary>
        /// Returns the major/minor version number for an assembly formatted as a string.
        /// </summary>
        private static string GetAssemblySoftwareVersion()
        {
            Version version = Assembly.GetCallingAssembly().GetName().Version;
            return Format("{0}.{1}", version.Major, version.Minor);
        }

        /// <summary>
        /// Returns the build/revision number for an assembly formatted as a string.
        /// </summary>
        private static string GetAssemblyBuildNumber()
        {
            Version version = Assembly.GetCallingAssembly().GetName().Version;
            return Format("{0}.{1}", version.Build, (version.MajorRevision << 16) + version.MinorRevision);
        }

        /// <summary>
        /// Formats a message using the default locale.
        /// </summary>
        private static string Format(string text, params object[] args)
        {
            return String.Format(CultureInfo.InvariantCulture, text, args);
        }

        public StatusCode OnCallMethod(object sender, MethodCallEventArgs e)
        {
            return StatusCodes.Good;
        }

        #endregion
    }
}
